src/Sylius/Bundle/CoreBundle/Controller/ProductTaxonController.php

```diff
@@ -56,35 +63,75 @@ public function updatePositionsAction(Request $request): Response
+        $this->validateCsrfProtection($request, $configuration);
         $this->isGrantedOr403($configuration, ResourceActions::UPDATE);
-        $productTaxons = $this->getParameterFromRequest($request, 'productTaxons');
+        $productTaxonsPositions = $request->request->all('productTaxons');
 
-        $this->validateCsrfProtection($request, $configuration);
+        if (!$this->shouldProductsPositionsBeUpdated($request, $productTaxonsPositions)) {
+            return $this->redirectHandler->redirectToReferer($configuration);
+        }
 
-        if ($this->shouldProductsPositionsBeUpdated($request, $productTaxons)) {
+        $maxPosition = $this->getMaxPosition($configuration);
+
+        try {
+            $this->updatePositions($productTaxonsPositions, $maxPosition);
+        } catch (\InvalidArgumentException $exception) {
             /** @var Session $session */
             $session = $request->getSession();
+            $session->getFlashBag()->add('error', $exception->getMessage());
 
-            foreach ($productTaxons as $id => $position) {
-                try {
-                    $this->updatePositions($position, $id);
-                } catch (\InvalidArgumentException $exception) {
-                    $session->getFlashBag()->add('error', $exception->getMessage());
+            return $this->redirectHandler->redirectToReferer($configuration);
+        }
 
-                    return $this->redirectHandler->redirectToReferer($configuration);
-                }
+
+        return $this->redirectHandler->redirectToReferer($configuration);
+    }
+
+    /** @param array<int, string> $productTaxonPositions */
+    private function updatePositions(array $productTaxonPositions, int $maxPosition): void
+    {
+        $positioner = $this->getPositioner();
+        $modifiedProductTaxons = [];
+
+        /** @var array<ProductTaxonInterface> $productTaxons */
+        $productTaxons = $this->repository->findBy(['id' => array_keys($productTaxonPositions)]);
+
+        foreach ($productTaxons as $productTaxon) {
+            $newProductTaxonPosition = $productTaxonPositions[$productTaxon->getId()];
+            if (!is_numeric($newProductTaxonPosition)) {
+                throw new \InvalidArgumentException(sprintf('The position "%s" is invalid.', $newProductTaxonPosition));
             }
 
+            $newProductTaxonPosition = (int) $newProductTaxonPosition;
+
+            if (!$positioner->hasPositionChanged($productTaxon, $newProductTaxonPosition)) {
+                continue;
+            }
+
+            $modifiedProductTaxons[] = [
+                'productTaxon' => $productTaxon,
+                'newPosition' => $newProductTaxonPosition,
+            ];
+        }
+
+        foreach ($modifiedProductTaxons as $modifiedProductTaxon) {
+            $positioner->updatePosition($modifiedProductTaxon['productTaxon'], $modifiedProductTaxon['newPosition'], $maxPosition);
             $this->manager->flush();
         }
```

Comment:
```suggestion
            $positioner->updatePosition($productTaxon, $newProductTaxonPosition, $maxPosition);
            $this->manager->flush();
        }
```
wdyt?

Reply:
It cannot be done within a single foreach. I guess we talked about it in some comment above on a daily 🤔.

If I update the position of a given item and flush it, the position of other elements is changed due to the doctrine event listener (TL;DR checking if the position has been changed stops working, as it sees every element's position has changed).